### ISSUE TITLE
[DOC] sktime forecasting 01 example

### ISSUE THREAD

**Issue by Data-drone** [alan-turing-institute/sktime#1212]
I am trying to follow the first example notebook but with my own dataset:

https://github.com/alan-turing-institute/sktime/blob/main/examples/01_forecasting.ipynb

I have two pandas series, train and test that have the schema:

index of type: `pandas.core.indexes.datetimes.DatetimeIndex`
series of type: `int64`

The series in train covers `2013-08-01` daily through to `2014-06-30`
The series in test covers `2014-07-01` daily through to `2016-06-30`

In order to generate predictions, I see that I need to create a _ForecastingHorizon_

so I create one as follows:

```
fh = ForecastingHorizon(
    pd.PeriodIndex(pd.date_range(test.index[0], 
                                 periods=len(ttest.index), 
                                 freq="D")), is_relative=False
)

```

as it seems like I can't just enter in the index of the testset.

I setup a _NaiveForecaster_ as described:

```
from sktime.forecasting.naive import NaiveForecaster

basic_forecaster =  NaiveForecaster(strategy="last")
forecaster = NaiveForecaster(strategy="last")
forecaster.fit(train)
```

but when I enter in the forecasting horizon to produce my predictions:
`forecaster.predict(fh)`

I get:

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-80-c84472c9f3f0> in <module>
----> 1 forecaster.predict(fh)

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_base.py in predict(self, fh, X, return_pred_int, alpha)
    156         # alpha = check_alpha(alpha)
    157 
--> 158         return self._predict(self.fh, X, return_pred_int=return_pred_int, alpha=alpha)
    159 
    160     def fit_predict(

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_sktime.py in _predict(self, fh, X, return_pred_int, alpha)
     77 
     78         # all values are out-of-sample
---> 79         if fh.is_all_out_of_sample(self.cutoff):
     80             return self._predict_fixed_cutoff(
     81                 fh.to_out_of_sample(self.cutoff), **kwargs

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in is_all_out_of_sample(self, cutoff)
    403             cutoff.
    404         """
--> 405         return sum(self._is_out_of_sample(cutoff)) == len(self)
    406 
    407     def to_indexer(self, cutoff=None, from_cutoff=True):

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in _is_out_of_sample(self, cutoff)
    385         """Get index location of out-of-sample values"""
    386         # return ~self._in_sample_idx(cutoff)
--> 387         return self.to_relative(cutoff).to_pandas() > 0
    388 
    389     def is_all_out_of_sample(self, cutoff=None):

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in to_relative(self, cutoff)
    236         else:
    237             absolute = self.to_pandas()
--> 238             _check_cutoff(cutoff, absolute)
    239 
    240             if isinstance(absolute, pd.DatetimeIndex):

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in _check_cutoff(cutoff, index)
    443 
    444     if isinstance(index, pd.PeriodIndex):
--> 445         assert isinstance(cutoff, pd.Period)
    446         assert index.freqstr == cutoff.freqstr
    447 

AssertionError: 
```
The forecasting horizon object seems to be valid? And since I have set an exact range I have set `is_relative` to false

**Comment by fkiraly** [comment]
I think your fh index must be of the same type as the index in your train/test set.

Your data set index seems to be `DatetimeIndex`, but your `fh` is a `PeriodIndex`, which is not the same, and there is no "obvious" coercion.

Is that the issue, possibly?

If so, this highlights that there should perhaps be a more informative error message.

**Comment by fkiraly** [comment]
Was it that, @Data-drone? Please confirm.
If not, would appreciate a reproducible example.

**Comment by Data-drone** [comment]
Hmm if I tried changing my fh index. So now it is:

```
fh = ForecastingHorizon(
    values=test.index, is_relative=False
)

```

where

test.index is a `DateTimeIndex`

but then I get:

```

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-129-20c4ea07743c> in <module>
      1 # stuck here for now
----> 2 forecaster.predict(fh)

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_base.py in predict(self, fh, X, return_pred_int, alpha)
    156         # alpha = check_alpha(alpha)
    157 
--> 158         return self._predict(self.fh, X, return_pred_int=return_pred_int, alpha=alpha)
    159 
    160     def fit_predict(

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_sktime.py in _predict(self, fh, X, return_pred_int, alpha)
     77 
     78         # all values are out-of-sample
---> 79         if fh.is_all_out_of_sample(self.cutoff):
     80             return self._predict_fixed_cutoff(
     81                 fh.to_out_of_sample(self.cutoff), **kwargs

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in is_all_out_of_sample(self, cutoff)
    403             cutoff.
    404         """
--> 405         return sum(self._is_out_of_sample(cutoff)) == len(self)
    406 
    407     def to_indexer(self, cutoff=None, from_cutoff=True):

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in _is_out_of_sample(self, cutoff)
    385         """Get index location of out-of-sample values"""
    386         # return ~self._in_sample_idx(cutoff)
--> 387         return self.to_relative(cutoff).to_pandas() > 0
    388 
    389     def is_all_out_of_sample(self, cutoff=None):

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in to_relative(self, cutoff)
    236         else:
    237             absolute = self.to_pandas()
--> 238             _check_cutoff(cutoff, absolute)
    239 
    240             if isinstance(absolute, pd.DatetimeIndex):

~/.local/lib/python3.8/site-packages/sktime/forecasting/base/_fh.py in _check_cutoff(cutoff, index)
    450 
    451         if not hasattr(cutoff, "freqstr") or cutoff.freqstr is None:
--> 452             raise AttributeError(
    453                 "The `freq` attribute of the time index is required, "
    454                 "but found: None. Please specify the `freq` argument "

AttributeError: The `freq` attribute of the time index is required, but found: None. Please specify the `freq` argument when setting the time index.

```

**Comment by fkiraly** [comment]
@Data-drone, thanks for bringing this up!

I looked into it, and it relates to two issues of "unintuitiveness" around `DatetimeIndex` in version 0.7.0:

* if `predict` is given an absolute horizon, it breaks, instead of converting into a `ForecastingHorizon`. This was highlighted by @flix6x in #1167, and has been addressed in the newest development version. So the *first* breakage point is resolved, just not in the version that's on `pip`.
* the second error relates to the convention tha `DatetimeIndex` must have a `freq` parameter set - yours is likely `None`. The reason is that without this most methods can't easily translate the dates into integer multiples of a period, internally. If you created the index yourself, you may want to set it explicitly or infer it, see the [`pandas` docs here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.html).

Looking at this second point, I agree that the message is ambiguous and the behaviour is unintuitive.

I would hence propose is the following behaviour in this case, i.e., if the index is `DatetimeIndex` and `freq` is `None` (not set), in `predict`:

* integer index is assumed on both train data and forecast horizon
* a warning message is displayed for forecasters that cannot handle non-equally spaced data, stating that the method is considering it equally spaced

@Data-drone, what do you think of this behaviour? @flix6x, what do you think (since this is closely related to your problem in #1167)?

@mloning, @aiwalter (authors of original `fh` logic), FYI, any comments?

**Comment by fkiraly** [comment]
PS, @Data-drone: the workaround for version 0.7.0 is to:

* construct the `fh` with `is_relative=False` before passing to `predict` - this will no longer be required in 0.7.1 onwards
* ensure `freq` is set in all `DatetimeIndex` of yours, and the same

**Comment by Data-drone** [comment]
How do I set freq with a `DatetimeIndex`?

**Comment by fkiraly** [comment]
Well, you can set it in the constructor when you create it, as outlined in the examples in the [pandas doc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.html).
So, for instance `myindex = DatetimeIndex(sthsth, freq="D")` for daily frequency.

You can also convert a `freq`-less pandas object using the [`asfreq` function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.asfreq.html).

If you post your code (doesn't need to be with the original data frames), happy to help you modifying it.

A quick hack would also be just to drop the index entirely, if it's equally spaced, using `reset_index(drop=True)` - you'd then have to calculate back to days.

**Comment by Data-drone** [comment]
so I am extracting my raw data from Spark and it arrives as a pandas object kind of like:

```
import datetime
import pandas as pd

tt = [datetime.date(2013,8,1), datetime.date(2013,8,2), datetime.date(2013,8,2)]
t2 = [0,1,2]

pd.DataFrame([x for x in zip(tt,t2)], columns=['ds', 'values'])

```

**Comment by fkiraly** [comment]
Hm, but this code doesn't create a `DatetimeIndex`? It just creates an index.
The reason is that non-pandas `datetime` can confuse the integration with `pandas` objects.

To ensure you get a `DatetimeIndex` with `freq` defined, you should preferably work with `pandas` objects and do sth like

```python

tt = pd.date_range("2013-08-01", "2013-08-05")
df = pd.DataFrame(t2, index=tt)
```

this will even automatically infer the `freq` to be `"D"`, which you can check by typing
```df.index```

**Comment by Data-drone** [comment]
Yeah I was converting it prior to feeding into sktime with:

```
train_sktime = pd.Series(data=train['y'].values, index=pd.to_datetime(train['ds']))
```
but figured maybe could save that step

**Comment by fkiraly** [comment]
>but figured maybe could save that step

well, there's no way for `sktime` to know whether you want to use the indices in the `ds` column or the actual data frame indices, so this would be a case of "reading your mind" which I don't think there is a good automation for.

But, on the other hand, I think the discussion around `freq` has highlighted some very useful insights, i.e., what to do if `sktime` sees a `DatetimeIndex` with no `freq` attribute. Guess we can make that more user friendly by automatically removing the index and interpreting it as equally spaced.

**Comment by Data-drone** [comment]
Okay so the workflow for me should be:

- Read in the data set from Spark
- Split the data into my train / validation dataset
      - The plan is to do sliding window and or expanding window cross validation with the train set then get final metrics on the 
         validation set

- Create the date_range objects by selecting `min()` and `max()` from my date column and create pandas `date_range` objects 
   with them

**Comment by Flix6x** [comment]
> [...] what to do if `sktime` sees a `DatetimeIndex` with no `freq` attribute. Guess we can make that more user friendly by automatically removing the index and interpreting it as equally spaced.

tldr; Being able to pass the freq explicitly would have my preference.

I am not exactly sure what you have in mind, but I imagine that ignoring the index and interpreting the data as equally spaced could effectively shift around data in a way that would result in an incorrect application of seasonal periodicity (for training data) or intended horizons (for testing data). In my experience, a `DatetimeIndex` with no `freq` attribute is often a time series with either some missing data points (missing an index) or duplicate data points (having equal indices), with some of them not even being missed (or used), depending on the model and the seasonal periodicity.

If you want to handle such cases without throwing back an informative exception to the user, I think the user should be able to pass the intended frequency to sktime explictly (even with priority over the freq from the DatetimeIndex) . You might then consider reindexing internally using `date_range` and the `min()` and `max()` of the original index (i.e. introducing NaN values for formerly missing indices, but I'm not sure what happens to formerly duplicate indices). Especially for training data, you could also consider imputing missing data points and weighing duplicate data points to avoid models tripping over them.

**Comment by Data-drone** [comment]
Good points will take that in mind. thanks for your assistance
